src/python/pants/backend/codegen/grpcio/grpcio_run.py

```diff
@@ -0,0 +1,84 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import functools
+import logging
+from collections import OrderedDict
+
+from pants.backend.codegen.grpcio.grpcio_prep import GrpcioPrep
+from pants.backend.codegen.grpcio.python_grpcio_library import PythonGrpcioLibrary
+from pants.backend.python.targets.python_library import PythonLibrary
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.task.simple_codegen_task import SimpleCodegenTask
+from pants.util.contextutil import pushd
+
+
+class GrpcioRun(SimpleCodegenTask):
+  """[XAIN] Task to compile protobuf"""
+
+  gentarget_type = PythonGrpcioLibrary
+  sources_globs = ('**/*',)
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(GrpcioRun, cls).prepare(options, round_manager)
+    round_manager.require_data(GrpcioPrep.Grpcio)
+
+  def execute_codegen(self, target, target_workdir):
+    grpcio = self.context.products.get_data(GrpcioPrep.Grpcio)
+
+    args = self.build_args(target, target_workdir)
+    logging.debug("Executing grpcio code generation with args: [{}]".format(args))
+
+    with pushd(get_buildroot()):
+      workunit_factory = functools.partial(self.context.new_workunit,
+                                           name='run-grpcio',
+                                           labels=[WorkUnitLabel.TOOL, WorkUnitLabel.LINT])
+      cmdline, exit_code = grpcio.run(workunit_factory, args)
+      if exit_code != 0:
+        raise TaskError('{} ... exited non-zero ({}).'.format(cmdline, exit_code),
+                        exit_code=exit_code)
+      logging.debug("Grpcio finished code generation")
+
+  def build_args(self, target, target_workdir):
+    python_out = '--python_out={0}'.format(target_workdir)
+    grpc_python_out = '--grpc_python_out={0}'.format(target_workdir)
+
+    args = [python_out, grpc_python_out]
+
+    sources = []
+    sources_by_base = self._calculate_sources(target)
+    for (base, source_list) in sources_by_base.items():
+      args.append('--proto_path={0}'.format(base))
+      sources.extend(source_list)
```

Comment:
Could you add an integration test that covers the case of building a grpcio target that depends on another grpcio target? I'm a little bit unsure whether adding all of the transitive sources is the correct thing to do here... it feels like it would result in codegening everything in the transitive graph for every target.

Reply:
I am working on it right now - probably I had this idea of 'collecting' sources from the previous approach - I am trying to rewrite it to use sources from selected target only. 

Also, will provide some test for that 👍

Reply:
This transitive targets selection is removed now, and still generated code seems to be valid. Even, if there is a dependency on `proto` files. I've updated examples for that situation and added 2 tests to cover this behaviour.